TT-related things from TR

grdddj opened the pull request:
As agreed, I have taken all the things relevant to `TT` from the big `TR`'s PR - https://github.com/trezor/trezor-firmware/pull/2610 - so that we can quickly review, merge them and build upon these.

The changes should be structured commit-by-commit, but with no guarantee that they are standalone.

mmilata commented on core/embed/rust/src/ui/component/text/layout.rs:
```diff
@@ -118,87 +144,76 @@ impl TextLayout {
         while !remaining_text.is_empty() {
+            let is_last_line = cursor.y + self.style.text_font.line_height() > self.bottom_y();
+            let line_ending_space = if is_last_line {
+                // TODO: find out the icon width
```
This.

grdddj replied:
Done in [b6fb917](https://github.com/trezor/trezor-firmware/pull/2803/commits/b6fb917d3111decae1eebac6d2a267c02e6b715d)

mmilata commented on core/embed/rust/src/ui/component/text/layout.rs:
```diff
@@ -118,87 +144,76 @@ impl TextLayout {
+            let is_last_line = cursor.y + self.style.text_font.line_height() > self.bottom_y();
+            let line_ending_space = if is_last_line {
+                // TODO: find out the icon width
+                let width = self.style.text_font.text_width(PREV_PAGE_ELLIPSIS);
```
We're at the end of the page though?

grdddj replied:
Fixed in [b6fb917](https://github.com/trezor/trezor-firmware/pull/2803/commits/b6fb917d3111decae1eebac6d2a267c02e6b715d)

mmilata commented on core/embed/rust/src/ui/component/text/layout.rs:
```diff
@@ -298,34 +324,81 @@ impl LayoutFit {
+    /// Ellipsis at the end of the page.
     fn ellipsis(&mut self, _cursor: Point, _layout: &TextLayout) {}
+    /// Ellipsis at the beginning of the page.
+    fn prev_page_ellipsis(&mut self, _cursor: Point, layout: &TextLayout) -> i16 {
```
I don't think this belongs here, sink isn't supposed to know anything about what the ellipsis looks like.

grdddj replied:
Reworked in [cb381ea](https://github.com/trezor/trezor-firmware/pull/2803/commits/cb381ea80ecfa737360f8fb855921c98467bd163)

mmilata reviewed (changes requested):
Added some feedback. Please also fix the UI tests so we can see the UI diff and code coverage. If it helps anything you can drop the testing-related changes from this PR as they are not really needed for the upcoming release.

grdddj commented on core/src/trezor/ui/layouts/tt_v2/__init__.py:
```diff
@@ -531,12 +503,15 @@ async def confirm_output(
     if title.startswith("CONFIRM "):
         title = title[len("CONFIRM ") :]
 
+    # TODO: include index to be consistent with TR?
```
CC @Hannsek

grdddj commented on core/src/apps/common/request_pin.py:
```diff
@@ -125,7 +119,7 @@ async def error_pin_invalid(ctx: Context) -> NoReturn:
         ctx,
         "warning_wrong_pin",
-        "The PIN you entered is invalid.",
+        "PIN you have entered is not valid.",
```
CC @Hannsek

Hannsek replied:
CC @b-irving

b-irving replied:
@Hannsek should include `The`

grdddj replied:
Included `The` in [e804557](https://github.com/trezor/trezor-firmware/pull/2803/commits/e804557217fb7a39ee5aa595857559697e94495a)

grdddj commented:
> Added some feedback. Please also fix the UI tests so we can see the UI diff and code coverage. If it helps anything you can drop the testing-related changes from this PR as they are not really needed for the upcoming release.

Thanks for the comments, I will address most of them later.

[7187eb8](https://github.com/trezor/trezor-firmware/pull/2803/commits/7187eb8500648a20f39933fccd241a165bbbc833) should fix the click-UI-tests

grdddj commented on core/src/apps/management/change_pin.py:
```diff
@@ -50,52 +54,53 @@ async def change_pin(ctx: Context, msg: ChangePin) -> Success:
+
     if msg.remove and has_pin:  # removing pin
-        return confirm_action(
+        return confirm_pin_action(
```
I am changing it back to `confirm_action` - and making the same change in `change_wipe_code.py`, so it looks the same

grdddj replied:
Done in one of the commits during rebase

grdddj commented on core/embed/rust/src/ui/mod.rs:
```diff
@@ -4,11 +4,12 @@ pub mod macros;
 pub mod animation;
 pub mod component;
 pub mod constant;
+pub mod debug;
```
I was unable to somehow "get it and print it" - as in our `nostd` environment we do not have functionality like `format!` to get the debug string and then print it.

grdddj replied:
`#[cfg(feature = "ui_debug")]` added in [ee38099](https://github.com/trezor/trezor-firmware/pull/2803/commits/ee38099f5b2bb7c513d95350d0e5bf9377814bc3)

grdddj commented on core/embed/rust/src/ui/macros.rs:
```diff
@@ -4,3 +4,35 @@ macro_rules! include_res {
+
+#[allow(unused_macros)] // Mostly for debugging purposes.
+/// Transforms bool into string slice. For example for printing.
+macro_rules! booltostr {
```
Deleted in [3c67f0d](https://github.com/trezor/trezor-firmware/pull/2803/commits/3c67f0dd15080dc81848d28a4a06fde61b98e37c)

grdddj commented on core/embed/rust/src/ui/macros.rs:
```diff
@@ -4,3 +4,35 @@ macro_rules! include_res {
+
+#[allow(unused_macros)] // Only used in TR so far.
+/// Concatenates arbitrary amount of slices into a String.
+macro_rules! build_string {
```
The benefit of this being a macro instead of a function is that one can use it without any import. Not sure what are the implications like binary size etc. Does it have any advantage of using a function instead?

grdddj commented on core/embed/rust/src/ui/component/text/formatted.rs:
```diff
@@ -1,284 +0,0 @@
-use core::{
```
Resurrected in rebase and force-push

grdddj commented on core/embed/rust/src/ui/component/text/layout.rs:
```diff
@@ -298,34 +324,81 @@ impl LayoutFit {
+        // Assume the current line can be drawn on from the cursor
+        // to the right side of the screen.
+
+        match layout.style.line_alignment {
```
Right, it was duplicate and currently even broken. Deleted in [bf8e1ff](https://github.com/trezor/trezor-firmware/pull/2803/commits/bf8e1ff1da3cefeb50a9a3f29ec06876cca0caa6)

grdddj commented on crypto/slip39_wordlist.h:
```diff
@@ -216,7 +41,7 @@ static const char* const slip39_wordlist[WORDS_COUNT] = {
 static const struct {
```
I did not have a reason to move it, but if we agree, will do it.

Failed crypto build should be solved by [bdad2d0](https://github.com/trezor/trezor-firmware/pull/2803/commits/bdad2d053bd08579a5bb483ffa108184ad2cb47c)

mmilata replied:
It's just that the arrays are quite related so I'd expect them in the same place, but it's a nitpick.

grdddj commented:
> If it helps anything you can drop the testing-related changes from this PR as they are not really needed for the upcoming release.

The tests are, unfortunately, needed here, as they reflect some of the UI changes like wording changes or some of the new screens.

---

What is interesting right now is that the `core device test` is green even when it failed one testcase - https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/3762985554 and the `unix UI change` is failing, even though I rebased on `master` and it should get the fix - https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/3762985591

---

The UI tests have one bug that I had encountered locally yesterday but did not yet manage to fix - https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3762985554/artifacts/test_ui_report/failed/TT-device_tests-bitcoin-test_signmessage.py::test_signmessage_pagination[utf_text].html

The last three characters of the message are not shown on the last page. The reason I found is that the last page with the content is placed differently than all the previous pages - it does not place it on the whole page from top to bottom, just on the minimal area where the text fits (and also seems to centralize it vertically). The issue seems that it does not count with the ellipsis "..." at the beginning, and so in this case it thinks it will fit on 3 lines, but in reality (with the extra "..."), it won't. As the higher component "knows"/thinks it is the last page, even though there are 3 more characters to output, these do not get shown on the next page, as there is no next page.

I am investigating how to integrate the "..." into the estimation of the height of the last screen.

![image](https://user-images.githubusercontent.com/42543243/218698699-e803adf3-9e7d-4cf1-8678-63ea7c5f6e33.png)

grdddj commented:
> ![image](https://user-images.githubusercontent.com/42543243/218698699-e803adf3-9e7d-4cf1-8678-63ea7c5f6e33.png)

Plot twist - the bug was in a little different place than I thought previously. The issue was caused by the trailing whitespace in the message (`"...ódy. Příšerně žluťoučký kůň úpěl ďábelské ódy. "`). This has caused a little different behavior when calling `TextLayout::layout_text` between  `layout.fit_text` and `layout.render_text` - just because of that trailing whitespace. 

When first calling `fit_text` to calculate the number of pages, the area it will fit on the page is unknown, therefore we will allocate the whole screen area for it. It found out that the text could be successfully fit on just three lines - the last "space" is not being printed, and therefore it fits the line just perfectly.

However, when then calling `render_text`, it already knows it has three lines to fit itself in, and when it reaches the last line and has the string `"el dabelske ody. "`, it would find out that this text will _not_ fit the whole line (based on the whole text length, including the final `" "`), and therefore would put an ellipsis at the end, not showing the last 3 (4) characters anywhere.

The fix seems to be always trimming the trailing whitespace from the string when checking whether it can fit the line completely.

![image](https://user-images.githubusercontent.com/42543243/218721839-05b48706-ac6b-4420-8132-5511495fa3af.png)


Should be solved by [b937dbc](https://github.com/trezor/trezor-firmware/pull/2803/commits/b937dbc3a899bb6420ec5093c640002ebcb39624)

grdddj commented:
Updated on the current situation, no more force-pushes will be done - ready for review. 

Centralized input-flows introduced in [e5d8eb4](https://github.com/trezor/trezor-firmware/pull/2803/commits/e5d8eb463455e929cf62465269c83752c269128b) and used in [a5feed9](https://github.com/trezor/trezor-firmware/pull/2803/commits/a5feed9e4fac70b9604dcacef38a22afeda8bcdf)

grdddj commented:
Rebased on the current `master`, together with introducing some Rust dead code deletion and scripts to enforce QSTRings and Rust API consistency

matejcik commented:
what is the reason for "crypto: SLIP39 refactor"?
(fwiw, the spec explicitly says that only the English wordlist is ever supported and there is not going to be any sort of localization: https://github.com/satoshilabs/slips/blob/master/slip-0039.md#localization . so even the rename to english is, well, not _incorrect_ per se but not in the intended direction)

grdddj commented:
> what is the reason for "crypto: SLIP39 refactor"? (fwiw, the spec explicitly says that only the English wordlist is ever supported and there is not going to be any sort of localization: https://github.com/satoshilabs/slips/blob/master/slip-0039.md#localization . so even the rename to english is, well, not _incorrect_ per se but not in the intended direction)

The only reason was to be able to export the wordlist into `Rust` in the same way as we export the BIP39 wordlist - so I have copied the code structure of BIP39. For the naming, we could drop the `english` suffix

mmilata commented on core/embed/rust/src/ui/debug.rs:
```diff
@@ -0,0 +1,219 @@
+// Debug {fn print(&self);}, so that the trait does
+// not need to be imported when using the
+// print() function. It suits the use-case of being quickly
+// able to use the print() for debugging and then delete it.
```
I'm not sure the ease of use justifies having to have every member of supported structs as `pub`. Let's keep it if you find it useful, but personally I'd prefer to avoid maintaining hand-written serialization functions. Wouldn't writing a `Tracer` impl that produces single line strings work just as well?

matejcik replied:
worth considering bringing in ufmt for the `uDebug` trait

grdddj replied:
I made these debug things a long time ago, and almost do not use them nowadays, so we can maybe freely delete them. 

> Wouldn't writing a Tracer impl that produces single line strings work just as well?

Will think about this

> worth considering bringing in ufmt for the uDebug trait

I don't know how I would do this

mmilata reviewed (commented):
Overall looking pretty good, good job with the inputflow centralization. I'm assuming the tracing changes will be superseded by matejcik's JSON thing?

matejcik commented on core/embed/rust/src/micropython/iter.rs:
```diff
@@ -21,15 +21,14 @@ impl IterBuf {
     }
 }
 
-pub struct Iter<'a> {
```
> Perhaps iter_buf should be transformed into PhantomData<'a IterBuf> instead of removing?

won't help because we actually need to mutably grab that particular IterBuf instance

matejcik commented:
added the JSON tracing in https://github.com/trezor/trezor-firmware/compare/matejcik/jsondebuglink?expand=1

this is still just a draft, no support on Python side nor in device tests

grdddj commented on core/Makefile:
```diff
@@ -352,3 +352,6 @@ coverage:  ## generate coverage report
 	vulture src src/_vulture_ignore.txt --exclude "messages.py,*/enums/*"
+
+qstr_check: ## check the consistency of QSTR definitions
+	./tools/qstr_checks.py
```
Something like going through the whole Rust codebase, finding all the used `MP_QSTR_XXX` values, and sorting them alphabetically? Could be viable.

One disadvantage it would have is no longer grouping them "logically", e.g. we have `// protobuf` at the top with 5 messages

mmilata replied:
yes; the logical grouping isn't really useful for anything

grdddj replied:
Done in [2d3a0aa](https://github.com/trezor/trezor-firmware/pull/2803/commits/2d3a0aa2bba3f73653c3c01e66258b82b1f15530)

mmilata replied:
Can you turn the script into mako template? Then it wouldn't need new Makefile targets.

grdddj replied:
Making it a template using current automation would mean adding the QSTR-generating logic into `common/tools/cointool.py::render`, which handles the `mako` templates.

That does not seem to me like a good place - `cointool.py` has nothing to do with `Rust`.

We might theoretically setup some other template automation more connected with `core`

mmilata replied:
You can put all the logic directly in the mako file (example: `all_modules.py.mako`), no?

grdddj replied:
Correct, very good idea. Done in [5fbf7ee](https://github.com/trezor/trezor-firmware/pull/2803/commits/5fbf7ee937f432a19269e9bd231aef6f27313845)

grdddj commented on core/Makefile:
```diff
@@ -128,11 +131,14 @@ test_emu_ui_record: ## record and hash screens for ui integration tests
 test_emu_accept_fixtures:  # accept UI fixtures from the last run of UI tests
 	../tests/update_fixtures.py
 
+test_emu_accept_fixtures_remove_missing:  # as above + remove all missing tests
```
`test_emu_accept_fixtures` is very useful locally when you run just a subset of tests and then want to update these in `fixtures.json` (without deleting those that did not run).

So I find both of these valuable, and (IMO) the point of `make` commands is to make our life easier via exposing useful/frequently run actions.

grdddj commented on core/embed/rust/src/ui/macros.rs:
```diff
@@ -4,3 +4,23 @@ macro_rules! include_res {
+    ($max:expr, $($string:expr),+) => {
+        {
+            let mut new_string = String::<$max>::new();
+            $(new_string.push_str($string).unwrap();)+
```
Done in [5b83266](https://github.com/trezor/trezor-firmware/pull/2803/commits/5b8326687607a438da6ea8cdb7b04e701479eefc)

grdddj commented on core/embed/rust/src/micropython/iter.rs:
```diff
@@ -21,15 +21,14 @@ impl IterBuf {
-pub struct Iter<'a> {
+pub struct Iter {
     iter: Obj,
-    iter_buf: &'a mut IterBuf,
```
Reverted in [5b83266](https://github.com/trezor/trezor-firmware/pull/2803/commits/5b8326687607a438da6ea8cdb7b04e701479eefc) + added `allow_dead_code` around it

grdddj commented on tests/common.py:
```diff
@@ -237,6 +249,20 @@ def input_flow():
+    debug.press_yes()
+
+
+def check_PIN_backoff_time(attempts: int, start: float) -> None:
```
Changed in [5b83266](https://github.com/trezor/trezor-firmware/pull/2803/commits/5b8326687607a438da6ea8cdb7b04e701479eefc)

grdddj commented on tests/common.py:
```diff
@@ -214,19 +220,25 @@ def input_flow():
 
     # check share
-    for _ in range(3):
+    for i in range(3):
```
Reverted in [5b83266](https://github.com/trezor/trezor-firmware/pull/2803/commits/5b8326687607a438da6ea8cdb7b04e701479eefc)

grdddj commented on tests/click_tests/test_passphrase_tt.py:
```diff
@@ -0,0 +1,286 @@
+assert DA_49_ADDRESS != DA_50_ADDRESS
+
+DA_51 = DA_50 + "d"
+DA_51_ADDRESS = "mg5L2i8HZKUvceK1sfmGHhE4gichFSsdvm"
```
Done in [5b83266](https://github.com/trezor/trezor-firmware/pull/2803/commits/5b8326687607a438da6ea8cdb7b04e701479eefc)

grdddj commented on core/Makefile:
```diff
@@ -88,13 +88,16 @@ test: ## run unit tests
 
 test_emu_multicore: ## run device tests using multiple cores
-	$(PYTEST) -n $(MULTICORE) $(TESTPATH)/device_tests $(TESTOPTS) --control-emulators --model=core --random-order-seed=$(shell echo $$RANDOM)
+	PYTEST_TIMEOUT=400 $(PYTEST) -n $(MULTICORE) $(TESTPATH)/device_tests $(TESTOPTS) \
```
Done in [5b83266](https://github.com/trezor/trezor-firmware/pull/2803/commits/5b8326687607a438da6ea8cdb7b04e701479eefc)

grdddj commented on tests/input_flows.py:
```diff
@@ -0,0 +1,1283 @@
+        raise NotImplementedError
+
+
+class InputFlowSetupDevicePINWIpeCode(InputFlowBase):
```
Allow me one last pushback before "decentralizing it" completely - as I started doing it, things started appearing that I really do not like. Mostly for slip39/bip39 input flows, the results just do not look good IMO.

I have done some of the "decentralization" in [c62d408](https://github.com/trezor/trezor-firmware/pull/2803/commits/c62d40849fc2a265d3a29d8c2ec562277819bcb4) and it uncovered for example these issues/unneeded complications:
- some input flows are used by more than one test-case, so they would need to be "centralized" anyway so we can reuse them
- the amount of imports that each device test needs to do for defining input flow is quite big - it needs to know a lot of details of how to click the button, how to swipe, how to assert ButtonRequestType etc.
- the result is then mixed of importing some InputFlows from common, and defining some locally

As one who spends a lot of time with device tests, having everything in `input_flows.py` is my preferred way, it can reuse all things/imports in this one file, and device tests do not have to care about any of this.

![image](https://user-images.githubusercontent.com/42543243/233948142-b3a9d180-d513-4f92-a22d-d9b704a70d5c.png)

The two-window view above (which everybody can do in VSCode by just clicking the InputFlowName) seems to me like an ideal situation - you see the high-level test-details on the left in a test file, while you can see/update the low-level input-flow-details on the right

grdddj commented:
I have cherry-picked the `Rust` JSON changes and added its support in `micropython` - together with the improvements in `debuglink` that make it more stable (taken from https://github.com/trezor/trezor-firmware/pull/2969, which I am closing therefore).

Also reverted the "decentralization" commit - as agreed, that will be a part of some future hackathon.

gat68 commented:
Enviado desde mi Samsung Mobile de Telcel
-------- Mensaje original --------De: Martin Milata ***@***.***> Fecha: 26/4/2023  4:01 a. m.  (GMT-06:00) A: trezor/trezor-firmware ***@***.***> CC: Subscribed ***@***.***> Asunto: Re: [trezor/trezor-firmware] TT-related things from TR (PR #2803) 
@mmilata commented on this pull request.



In core/Makefile:
> @@ -352,3 +352,6 @@ coverage:  ## generate coverage report
 
 unused:  ## find unused micropython code
 	vulture src src/_vulture_ignore.txt --exclude "messages.py,*/enums/*"
+
+qstr_check: ## check the consistency of QSTR definitions
+	./tools/qstr_checks.py

You can put all the logic directly in the mako file (example: all_modules.py.mako), no?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

matejcik commented on tests/click_tests/test_reset_slip39_advanced.py:
```diff
@@ -15,164 +15,113 @@
 @pytest.mark.setup_client(uninitialized=True)
 @pytest.mark.slow
-@with_mock_urandom
 def test_reset_slip39_advanced_16of16groups_16of16shares(
```
huh so the cases were identical except for the parameters? let's parametrize the single one then (we can add marks to individual parameters via `pytest.param` to indicate one is slow and the other is not)

matejcik replied:
> Would be still possible with parametrization, but current possibility of `-k 1of1` is easier

`pytest.param(..., id="1of1")` ?

grdddj replied:
Done in [b0522bb](https://github.com/trezor/trezor-firmware/pull/2803/commits/b0522bbd7ee7a16be1fce9bf3c38f360f6879bdd)

matejcik commented on tests/click_tests/test_reset_slip39_basic.py:
```diff
@@ -15,167 +15,109 @@
+
+@pytest.mark.setup_client(uninitialized=True)
+def test_reset_slip39_basic_16of16(device_handler: "BackgroundDeviceHandler"):
+    _slip39_basic_reset(device_handler, 16, 16)
```
dtto for parametrization

grdddj commented on core/embed/rust/src/ui/geometry.rs:
```diff
@@ -414,6 +430,16 @@ impl Rect {
+    pub fn corner_points(&self) -> [Point; 4] {
+        [
+            self.top_left(),
+            self.top_right() - Offset::x(1),
```
The reason is the difference between "point" and "edge" - just `top_right` is the right top edge of the `Rect`, but to get the point on the top right, one needs to subtract one pixel on the `x` - similarly for the other two corners.

I had to draw it myself to find this out, because without the Offsets it just does not work as expected

mmilata replied:
Sounds like `top_right`, `bottom_right`, `bottom_left` should be fixed as they return points outside the rectangle?

grdddj replied:
I have "fixed" the `top_right`, `bottom_right` and `bottom_left` and _all_ the UI tests have changes because of this - things like text centering in buttons or scrollbar placing is off because of it.

@mmilata are you OK with this, that probably you will need to fix the centering and other broken stuff on TT?

grdddj replied:
Done in [b3bf5f0](https://github.com/trezor/trezor-firmware/pull/2803/commits/b3bf5f00d6d3ee4ada60ea48769566ea17d61ba1) - you might want to wait for CI to finish the UI tests and see the diff

mmilata replied:
If it changes all UI fixtures and needs some additional work then let's do it outside of this PR.

grdddj replied:
Right, so I will revert the changes from this branch

grdddj replied:
Reverted in [dc60f22](https://github.com/trezor/trezor-firmware/pull/2803/commits/dc60f22f1e2c7ae511a20b99b397e13d5511225f)

grdddj replied:
Removed TODO in [fe85610](https://github.com/trezor/trezor-firmware/pull/2803/commits/fe85610ffe0085843d8fc2de4adf6a78c52e4e82)

grdddj commented on python/src/trezorlib/debuglink.py:
```diff
@@ -245,19 +332,30 @@ def state(self) -> messages.DebugLinkState:
+        return LayoutContent(self.state().json_content or "")
+
+    def wait_layout(self, wait_for_external_change: bool = False) -> LayoutContent:
+        # Next layout change will be caused by external event
```
This is rather a synchronization mechanism - when there is a change that we did not do directly by debug events - device being locked, running device_handler.run() etc., it will wait for this external change

matejcik replied:
no, what I meant is, why would you ever call `wait_layout(external_event=False)`. it kind of seems to me that, given the availability of `wait=True` on debuglink decisions, we could declare `wait_layout()` to mean "wait for a change caused externally" in all cases.

could that work or are there uses that are not covered?

grdddj commented on tests/click_tests/test_autolock.py:
```diff
@@ -92,17 +91,15 @@ def test_autolock_interrupts_signing(device_handler: "BackgroundDeviceHandler"):
 
     layout = debug.click(buttons.OK, wait=True)
-    assert "Total amount: 0.0039 BTC" in layout.get_content()
+    assert "0.0038 BTC" in layout.text_content()
```
It is because the first screen which we assert says - you are sending 0.38, and only a later screen says - you are sending 0,39 including 0.01 fee. This transaction-flow change is probably already in `master`, it is not rebased - my guess?

grdddj commented on tests/click_tests/test_passphrase_tt.py:
```diff
@@ -0,0 +1,297 @@
+        before = debug.read_layout().passphrase()
+    for char in passphrase:
+        debug.wait_layout()
+        press_char(debug, char)
```
We cannot, it sometimes "randomly" fails, when it clicks on the wrong time, e.g.
```
E             - Y@14lw%p)JN@f54MYvys@zj'g
E             ?   ^    ^            ^   ^
E             + Y@ 4lw%P)JN@f54MYvysA:zj'G
```

Good thing about the new `debuglink` is that now we can wait literally all the time without it causing freezes. Yeah, that has the disadvantage that we are maybe waiting somewhere it is not necessary, but that seems to be better than troubleshooting failures from not waiting enough.

matejcik replied:
that means that there is still some sort of instability. i would like to investigate that.
when you say "randomly", is it deterministic in some way (same corruption of same passphrase) or is it actually timing related?
is there a problem, for example, if you always stay on one screen (eg. all-lowercase)? or maybe if you are entering two  chars from the same key?

grdddj replied:
When running for a second time, another test failed, but all others have passed.

```
E           AssertionError: assert 'a       b   ...    e       f' == 'a       b   ...    e       f'
E             - a       b       c       d       e       f
E             ?                                -
E             + a       b       c       d      e       f
```

The third attempt was the same as the second - and therefore it seems that it has "issues" when entering the same key (space) multiple times in a row.

Feel free to try - `pytest ../tests/click_tests/test_passphrase_tt.py` (after I will push a fix to json trace which is broken now)

grdddj replied:
Sorry, it was my mistake, I was "fixing" the JSON trace (32 byte limit). Now it all works even without those `wait_layout` around - [61c0197](https://github.com/trezor/trezor-firmware/pull/2803/commits/61c019713cba6ddef0900b5f45b50645de4288dd)

grdddj commented on tests/click_tests/test_pin.py:
```diff
@@ -0,0 +1,283 @@
+) -> Generator["DebugLink", None, None]:
+    debug = device_handler.debuglink()
+    # So that random digits are always the same
+    debug.reseed(0)
```
Yeah, `setup_client` does it, but _only_ for UI tests, when you run simple device-tests, it is being random all the time (I was confused for a long time why I am seeing different seed words etc. in normal tests).

This is probably a leftover, will delete it after trying it works without it

grdddj replied:
Deleted in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on core/embed/rust/src/ui/macros.rs:
```diff
@@ -4,3 +4,23 @@ macro_rules! include_res {
+    }
+}
+
+#[allow(unused_macros)] // Mostly for debugging purposes.
```
Marked `ui_debug`-only in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on core/embed/rust/src/strutil.rs:
```diff
@@ -9,3 +9,34 @@ pub fn hexlify(data: &[u8], buffer: &mut [u8]) {
     }
 }
+
+pub fn format_i64(num: i64, buffer: &mut [u8]) -> Option<&str> {
```
Removed and replaced in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on core/src/storage/debug.py:
```diff
@@ -7,10 +7,26 @@
+        global debug_events
+
+        debug_events = DebugEvents()
+
```
Changed in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on core/src/trezor/ui/layouts/tt_v2/__init__.py:
```diff
@@ -60,44 +56,51 @@ def _paint(self) -> None:
-            return result
+            json_str = " ".join(rust_result)
+            return json_str
 
```
Done like this in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on core/src/trezor/wire/__init__.py:
```diff
@@ -466,7 +466,13 @@ def failure(exc: BaseException) -> Failure:
+        # NOTE: when receiving generic `FirmwareError` on non-debug build,
+        # change the `if __debug__` to `if True` to get the full error message.
+        if __debug__:
+            message = f"Firmware error - {exc}"
```
Done in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on tests/device_handler.py:
```diff
@@ -47,6 +47,11 @@ def run(self, function, *args, **kwargs) -> None:
             raise RuntimeError("Wait for previous task first")
         self.task = self._pool.submit(function, self.client, *args, **kwargs)
 
+    def run_and_wait_for_ui(self, function, *args, **kwargs) -> None:
```
Renamed and inlined in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on tests/click_tests/test_passphrase_tt.py:
```diff
@@ -0,0 +1,297 @@
+    if target_index > current_index:
+        for _ in range(target_index - current_index):
+            debug.swipe_left()
+            time.sleep(0.1)
```
Done in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented on tests/click_tests/test_passphrase_tt.py:
```diff
@@ -0,0 +1,297 @@
+        )
+        for _ in range(12):
+            delete_char(debug)
+            time.sleep(0.1)
```
We do not actually, it is probably a leftover from previous state, when the wait=True was not so reliable. Deleted in [6207744](https://github.com/trezor/trezor-firmware/pull/2803/commits/6207744f056f309254a9dc447711b31f43952a7e)

grdddj commented:
Rebased on current `master`.

[6f62629](https://github.com/trezor/trezor-firmware/pull/2803/commits/6f626291f219f8bd2d282b7121e5b83b58969c69) fixes https://github.com/trezor/trezor-firmware/issues/2977 - this was a good moment to create a click test for it and fix it

Also, I reintroduced the `debug.reseed(0)` in `PIN` click tests, as with `device_handler` this was somehow not being done

mmilata reviewed (commented):
Some more comments. Haven't reviewed changes in `core/src/` very deeply. Passphrase entry fix lgtm.

Good job with the json tracing @matejcik!

grdddj commented on core/embed/rust/src/ui/display/mod.rs:
```diff
@@ -803,7 +829,13 @@ pub fn dotted_line(start: Point, width: i16, color: Color) {
 
-pub fn text(baseline: Point, text: &str, font: Font, fg_color: Color, bg_color: Color) {
+/// Paints a pixel with a specific color on a given point.
+pub fn paint_point(point: &Point, color: Color) {
```
Naive question, what are the benefits of using `set_window & pixeldata` compared to this? It will probably be quicker, but for smaller things like rectangle corners (the only place where it is used) this one seems more simple.

mmilata replied:
Yes, it's mostly about speed, and that I wouldn't want encourage future people to use `paint_point` as it's the wrong thing to do in most cases. Rectangle corners are fine, I would just maybe use 4x `rect_fill()` directly there, but it's a nitpick.

matejcik replied:
> I wouldn't want encourage future people to use paint_point as it's the wrong thing to do in most cases

hard agree, and we can use `display::bar` in the `rect_fill_corners` directly

grdddj commented on core/embed/rust/src/ui/display/loader/rectangular.rs:
```diff
@@ -1,3 +1,5 @@
+#![allow(dead_code)] // clippy marks these unused, as they are only for TR
```
I cannot figure out how to do this properly. When I added the `clippy` for both models, it complains that `determinate` and `indeterminate` are reimported.

I have also dropped `#![cfg_attr(not(feature = "clippy"), allow(dead_code))]` in `lib.rs` and replaced it just with `#![allow(dead_code)]`, as there were a lot of places clippy found some false positives. It was that before, and if we just occasionally comment it and run `clippy`, we can see if something is really not used.

grdddj commented on core/embed/rust/src/ui/display/color.rs:
```diff
@@ -0,0 +1,110 @@
+            ($n * 256.0) as u16
+        }
+    };
+}
```
Deleted in [b3bf5f0](https://github.com/trezor/trezor-firmware/pull/2803/commits/b3bf5f00d6d3ee4ada60ea48769566ea17d61ba1)

grdddj commented on core/embed/rust/src/ui/model_tt/component/page.rs:
```diff
@@ -495,8 +494,10 @@ where
 
 #[cfg(test)]
 mod tests {
+    extern crate serde_json;
```
Switched to `use` instead of `extern create` in [b3bf5f0](https://github.com/trezor/trezor-firmware/pull/2803/commits/b3bf5f00d6d3ee4ada60ea48769566ea17d61ba1)

grdddj commented on tests/device_tests/test_sdcard.py:
```diff
@@ -21,7 +21,7 @@
 from trezorlib.messages import SdProtectOperationType as Op
 
-pytestmark = pytest.mark.skip_t1
+pytestmark = [pytest.mark.skip_t1, pytest.mark.sd_card]
```
RIght, I did not see it below. Deleted in [b3bf5f0](https://github.com/trezor/trezor-firmware/pull/2803/commits/b3bf5f00d6d3ee4ada60ea48769566ea17d61ba1)

grdddj commented on core/embed/rust/src/ui/model_tt/component/keyboard/pin.rs:
```diff
@@ -289,7 +288,7 @@ struct PinDots {
     style: TextStyle,
     digits: String<MAX_LENGTH>,
-    display_digits: bool,
+    pub display_digits: bool,
```
Deleted in [b3bf5f0](https://github.com/trezor/trezor-firmware/pull/2803/commits/b3bf5f00d6d3ee4ada60ea48769566ea17d61ba1)

grdddj commented on core/src/apps/common/request_pin.py:
```diff
@@ -146,7 +140,7 @@ async def error_pin_invalid(ctx: Context) -> NoReturn:
         ctx,
         "warning_wrong_pin",
-        "The PIN you entered is invalid.",
+        "The PIN you have entered is not valid.",
```
Cannot find it in `Figma` now, but it comes from the cooperation with @Hannsek 

Yes, it will change some TT screens, it is already accounted for in UI tests - when CI will be working, there will be master-diff

matejcik commented on core/embed/rust/src/micropython/iter.rs:
```diff
@@ -45,14 +46,6 @@ impl<'a> Iter<'a> {
         })
     }
-
-    pub fn error(&self) -> Option<Error> {
```
returning `Obj` and including the possible iterator exception confuses the iterator content with the iteration process -- if we ever need fallible iteration, the code to handle this distinction would be ugly

returning a `Result` would mean we still need to iterate like this:
```
for item in iter {
    let item = item?;
    // ...
}
```
and instead we could do something like:
```
while let Some(item) = iter.try_next()? {
   // ...
}
```
for a very weird signature `try_next() -> Result<Option<Obj>, Error>`

matejcik commented on core/tools/rust_api_models_unification.py:
```diff
@@ -0,0 +1,85 @@
+"""
+from __future__ import annotations
+
+import re
```
i suspect ast won't show you comments though? in that case you would need to use something like libcst
and maybe it's overkill for this purpose anyway

grdddj commented on core/embed/rust/src/ui/model_tr/theme.rs:
```diff
@@ -6,6 +6,9 @@ use crate::ui::{
 use super::component::{ButtonStyle, ButtonStyleSheet};
 
+// Typical backlight values.
+pub const BACKLIGHT_NORMAL: u16 = 150;
```
It is a `TR` code, I would move this into the `TR` PR and investigate it there. Is the statement true also for `TT`?

mmilata replied:
Ok. It is not true for TT, as it has LCD display with a backlight. TR has OLED display which generally don't have backlight, even though the previous board version had `set_backlight` call which did something (set brightness?). For current version `set_backlight` does nothing.

grdddj commented on core/embed/rust/src/ui/component/text/layout.rs:
```diff
@@ -70,6 +69,10 @@ pub struct TextStyle {
+    /// How many pixels to leave between the text and ellipsis icon.
+    pub ellipsis_icon_margin: i16,
+    /// How many pixels to leave between the prev_page icon and the text.
+    pub prev_page_icon_margin: i16,
```
Done in [cc85a29](https://github.com/trezor/trezor-firmware/pull/2803/commits/cc85a299ba19980eec841b50d4d9fe85a6325fd0)

grdddj commented on core/embed/rust/src/ui/model_tt/layout.rs:
```diff
@@ -862,19 +839,29 @@ extern "C" fn new_confirm_modify_fee(n_args: usize, args: *const Obj, kwargs: *m
-        ]);
+        let mut paragraphs_vec = ParagraphVecShort::new();
+        paragraphs_vec
+            .add(Paragraph::new(&theme::TEXT_BOLD, description.into()))
```
Here we change `TEXT_NORMAL` to `TEXT_BOLD` - is it wanted?

mmilata replied:
Figma has `TEXT_NORMAL`.

grdddj commented on core/embed/rust/src/strutil.rs:
```diff
@@ -9,3 +9,34 @@ pub fn hexlify(data: &[u8], buffer: &mut [u8]) {
 }
+
+pub fn format_i64(num: i64, buffer: &mut [u8]) -> Option<&str> {
+    let mut i = 0;
```
Cannot we theoretically use `inttostr!` instead of this? It is not very practical that calletwe need to supply the buffer every time.

We can call `inttostr!` from anywhere, we would probably need to make it return `Option`/`Result` because it could fail.

matejcik replied:
for now, we want to avoid creating adhoc Strings as much as we can

grdddj commented on core/src/apps/management/change_pin.py:
```diff
@@ -28,7 +28,11 @@ async def change_pin(ctx: Context, msg: ChangePin) -> Success:
+    if msg.remove:
+        prompt = "Enter PIN"
+    else:
+        prompt = "Enter old PIN"
```
This longer text is colliding with `15 attempts left` - cannot find the UI screenshots now

grdddj replied:
![image](https://user-images.githubusercontent.com/42543243/236181682-427785dd-fa92-4c9e-a6ab-3aff483b23b1.png)

I guess I will remove the `old`, so it is as before

Hannsek replied:
Can [this](https://github.com/trezor/trezor-firmware/issues/2996) be done together with this PR, or is it something else?

grdddj replied:
> Can https://github.com/trezor/trezor-firmware/issues/2996 be done together with this PR, or is it something else?

I would do it in a separate one

grdddj commented on core/embed/rust/src/ui/display/font.rs:
```diff
@@ -0,0 +1,227 @@
+        display::text_width(text, self.into())
+    }
+
+    /// Width of the text that is visible, not including the space
+    /// after the last character.
+    pub fn visible_text_width(self, text: &str) -> i16 {
+        let text_minus_one = &text[..text.len() - 1];
+
+        let last_char = unwrap!(text.chars().last());
+        let last_char_glyph = unwrap!(self.get_glyph(last_char as u8));
+
+        self.text_width(text_minus_one) + last_char_glyph.width
+    }
```
This is not used in `TT`, only introduced for `TR`. And also it is outdated here, this version does not count with bearing of the first character. Current `TR` situation below:

I guess I will remove it from this PR.

```rs
    /// Width of the text that is visible.
    /// Not including the spaces before the first and after the last character.
    pub fn visible_text_width(self, text: &str) -> i16 {
        match text.len() {
            0 => 0,
            1 => {
                let char = unwrap!(text.chars().last());
                let char_glyph = unwrap!(self.get_glyph(char as u8));
                char_glyph.width
            }
            _ => {
                let first_char = unwrap!(text.chars().next());
                let first_char_glyph = unwrap!(self.get_glyph(first_char as u8));
                let first_char_visible_width = first_char_glyph.adv - first_char_glyph.bearing_x;

                let middle_chars = &text[1..text.len() - 1];
                let middle_chars_width = self.text_width(middle_chars);

                let last_char = unwrap!(text.chars().last());
                let last_char_glyph = unwrap!(self.get_glyph(last_char as u8));
                let last_char_visible_width = last_char_glyph.width + last_char_glyph.bearing_x;

                first_char_visible_width + middle_chars_width + last_char_visible_width
            }
        }
    }
```

grdddj replied:
Deleted the function in [fe85610](https://github.com/trezor/trezor-firmware/pull/2803/commits/fe85610ffe0085843d8fc2de4adf6a78c52e4e82)

grdddj commented on core/embed/rust/src/trezorhal/bip39.rs:
```diff
@@ -1,48 +1,18 @@
-}
+use super::{ffi, wordlist::Wordlist};
 
 pub fn complete_word(prefix: &str) -> Option<&'static str> {
```
These three functions are not used in `slip39`, and `ffi::mnemonic_word_completion_mask` is specific for `bip39`.

I guess we could move them to `Wordlist` and prepend `bip39_` to their names?

```rs
pub fn complete_word(prefix: &str) -> Option<&'static str> 

pub fn options_num(prefix: &str) -> Option<usize>

pub fn word_completion_mask(prefix: &str) -> u32 
```

matejcik replied:
ah, okay, i see the problem now. let's keep this as-is and possibly keep in mind for some sort of refactor in the future.

grdddj commented on core/embed/rust/src/trezorhal/wordlist.rs:
```diff
@@ -0,0 +1,199 @@
+
+    /// Returns all possible letters that form a valid word together with some
+    /// prefix. Alphabetically sorted.
+    pub fn get_available_letters(&self, prefix: &str) -> impl Iterator<Item = char> {
```
Yes, the return value could be changed, we then do not need to `collect` it every time.

WIll write a note to consider this in TR's PR

grdddj commented on core/embed/rust/src/ui/display/font.rs:
```diff
@@ -0,0 +1,227 @@
+        if gl_data.is_null() {
+            return None;
+        }
+        unsafe { Some(Glyph::load(gl_data)) }
```
Done in [fe85610](https://github.com/trezor/trezor-firmware/pull/2803/commits/fe85610ffe0085843d8fc2de4adf6a78c52e4e82)

grdddj commented:
Force-pushed with cleaned-up commits

matejcik commented:
some checks are failing, see https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/858168834

there are still unresolved comments. I closed what I knew is done; @mmilata please check yours
@grdddj there's still the thing about return type of slip39 next letters -- maybe move the whole function over to the TR PR?

grdddj commented:
> some checks are failing, see https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/pipelines/858168834
> 

I am fixing all the failing tests and making sure the master diff is alright.

> @grdddj there's still the thing about return type of slip39 next letters -- maybe move the whole function over to the TR PR?
>

Right, I will delete the function (and its tests) from this PR.

I will push some fixups after all will be fixed

matejcik commented:
not sure if this got here, there are two issues that we can solve directly:
* in the "show error" dialog/popup, the order of title / description is flipped, so **The pin you have entered is not at all valid** (wrong pin) instead of **Wrong PIN** (the pin you have ...)
* the new screen about failed PIN should be using this "show error" dialog instead of confirm_action

@grdddj if this is not done yet, please make these changes and update the PR so we can merge

grdddj commented:
> not sure if this got here, there are two issues that we can solve directly:
> 
>     * in the "show error" dialog/popup, the order of title / description is flipped, so **The pin you have entered is not at all valid** (wrong pin) instead of **Wrong PIN** (the pin you have ...)
> 
>     * the new screen about failed PIN should be using this "show error" dialog instead of confirm_action
> 
> 
> @grdddj if this is not done yet, please make these changes and update the PR so we can merge

Right, I am reflecting these changes in tests now, I will force-push clean git history after it

grdddj commented:
Rebased on `master` and squashed all the fixups.

[4d6434c](https://github.com/trezor/trezor-firmware/pull/2803/commits/4d6434c221bdcdeb361df5edfcb34b4397848536) contains the error screens corrections.

![image](https://github.com/trezor/trezor-firmware/assets/42543243/d28a0215-3469-41f6-aee6-06012888f941)

Some considerations from these last changes:
- I tried to do it without `button_requests`, but then tests would not be working well (cannot click TRY AGAIN without knowing there is a new button-screen)
- I have left the ALL CAPS titles, maybe we want to make it Lower - as with `Invalid PIN`
- it is currently not possible to cancel from these screens, I thought it is better like this, and with the cancel button, the `TRY AGAIN` text almost did not fit the right button (this has consequences for tests, that cannot cancel on this page, need to press yes and cancel on the next page)

Hannsek commented:
Please change the font to lower, no all caps.

Thing I would revert: 

<img width="246" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/935a7af9-d4c2-4c7a-90f5-5e65bb9c2e2b">
<img width="244" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/43d6c6cf-fe68-4888-838e-0ee5ba5b9d42">
<img width="244" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/253fff5b-da81-4cbc-8504-1590785d5b3b">
<img width="245" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/b3325f8c-93b3-4478-a40d-7928b3b5f687">
<img width="248" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/649e9a19-53ce-4391-98a6-23308fdfc4ff">
<img width="248" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/a57dc045-f7de-45b3-9899-3297980b853b">
<img width="246" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/dc92a281-4efb-466d-b802-e5b1f161865f">
<img width="251" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/e6e235d9-5733-4922-b88b-d3ebdd0093ef">
<img width="247" alt="image" src="https://github.com/trezor/trezor-firmware/assets/57008159/a2a34d56-303c-4b5b-bc78-1f5b005994cc">


We are working on a quite big copy changes in fw. For that I will create new issue(s).

grdddj commented:
> Please change the font to lower, no all caps.

Done in the force-push.

All other changes will be done in a separate PR.

mmilata commented on core/embed/rust/src/trezorhal/wordlist.rs:
```diff
@@ -0,0 +1,161 @@
+    Ordering::Equal
+}
+
+unsafe fn from_utf8_unchecked<'a>(word: *const cty::c_char) -> &'a str {
```
Unsafe functions should have a docstring documenting their safety assumptions. This function can be made to return any lifetime which should probably be also documented.

grdddj commented:
CI is alright, apart from one failed build (most probably CI issues) - https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/4260222919

UI master diff - https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/4260223172/artifacts/master_diff/differing_screens.html

bosomt commented:
**QA OK**

**Info:**
 - Suite version: desktop 23.6.1 (bddd516378beadfcff94523e7c94834ba14f147f)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/23.6.1 Chrome/112.0.5615.183 Electron/24.3.1 Safari/537.36
 - OS: MacIntel
 - Screen: 1440x900
 - Device: model T 2.6.1 Universal (revision 089ff768229f543c89927ff34c4d45d830a2442f)
 - Transport: BridgeTransport 2.0.33
